include/oneapi/mkl/dft/detail/types_impl.hpp

```diff
@@ -39,6 +39,60 @@ struct precision_t {
 };
 
 enum class domain { REAL, COMPLEX };
+
+// Forward declarations
+template <precision prec, domain dom>
+class commit_impl;
+
+template <precision prec, domain dom>
+class descriptor;
+
+template <class... T>
+constexpr bool always_false = false;
+
+template <typename descriptor_type>
+struct descriptor_info {
+    static_assert(always_false<descriptor_type>, "Not a valid descriptor type");
```

Comment:
Yes, but won't there be an error anyway when `scalar_type`, `forward_type` and `backward_type` aren't defined? I think we should this and the `always_false` since `descriptor_info` will only ever be oneMKL developer-facing.

Reply:
There will be, but this is a good error in my opinion. e.g. I updated on of the tests to use `descriptor_t&` as the first template argument and part of the output was 
```
finlay@devpc-c310:~/perf/oneMKL/build$ ninja | head
[1/4] Building CXX object tests/unit_tests/dft/source/CMakeFiles/dft_source_rt.dir/compute_tests.cpp.o
FAILED: tests/unit_tests/dft/source/CMakeFiles/dft_source_rt.dir/compute_tests.cpp.o 
/home/finlay/intel/oneapi/compiler/2023.1.0/linux/bin/icpx  -I/home/finlay/perf/oneMKL/tests/unit_tests/dft/source/../include -I/home/finlay/perf/oneMKL/tests/unit_tests/dft/source/../../include -I/home/finlay/perf/oneMKL/include -I/home/finlay/perf/oneMKL/deps/googletest/include -I/home/finlay/perf/oneMKL/build/bin -O3 -DNDEBUG -std=gnu++17 -DCALL_RT_API -DNOMINMAX -fsycl -Wall -Wextra -Wshadow -Wconversion -Wpedantic -MD -MT tests/unit_tests/dft/source/CMakeFiles/dft_source_rt.dir/compute_tests.cpp.o -MF tests/unit_tests/dft/source/CMakeFiles/dft_source_rt.dir/compute_tests.cpp.o.d -o tests/unit_tests/dft/source/CMakeFiles/dft_source_rt.dir/compute_tests.cpp.o -c /home/finlay/perf/oneMKL/tests/unit_tests/dft/source/compute_tests.cpp
In file included from /home/finlay/perf/oneMKL/tests/unit_tests/dft/source/compute_tests.cpp:29:
In file included from /home/finlay/perf/oneMKL/tests/unit_tests/dft/source/../../include/test_helper.hpp:33:
In file included from /home/finlay/perf/oneMKL/include/oneapi/mkl.hpp:26:
In file included from /home/finlay/perf/oneMKL/include/oneapi/mkl/dft.hpp:33:
In file included from /home/finlay/perf/oneMKL/include/oneapi/mkl/dft/detail/dft_loader.hpp:30:
/home/finlay/perf/oneMKL/include/oneapi/mkl/dft/detail/types_impl.hpp:55:5: error: static assertion failed due to requirement 'always_false<oneapi::mkl::dft::detail::descriptor<oneapi::mkl::dft::detail::precision::SINGLE, oneapi::mkl::dft::detail::domain::COMPLEX> &>': Not a valid descriptor type
    static_assert(always_false<descriptor_type>, "Not a valid descriptor type");
```
potentially I could replace this with a static assert at the user-facing call site though.

Reply:
I'd like to keep this. When the user specifies the wrong template argument for the descriptor, it will give them a more plain error.